### ISSUE TITLE
SCSS maps are too eager to split onto multiple lines

### ISSUE THREAD

**Issue by BPScott** [prettier/prettier#4840]
Currently formatting SCSS map/list literals always splits maps that could fit into the print-width out over multiple lines.

e.g. 

```
$point: (x: 123);
```

always gets reformatted to

```
$point: (
  x: 123
);
```

I feel that if a map declaration can fit on a single line then it should be allowed to do so. This is what the babylon parser allows with with JS objects.

For reference [this is a playground like to similar constructs in JS](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAyjAhvAWwRgwEsBnDcgCwgCcSEIBXAc2oxggwDNSTSsbtDgYANoLgAdKJCjkSADwwBeDMADWcAJ5IMAcgBuuMczj6ANBi3aATEn3k4cgCYA1E2cvWdAZj36MNSkdO6e5gC+ANwy6BgA8gBKGATMYjCk4pJklHxQojz0VLQMGBAARgBWzjCUXBAycgoYymrAMhg+ugbGpuYWHV32js7QYX2Wgzb+BkEh414y0TKxmDj4cESwOcX0jFAs7JzcfAJCZflZ+Y3Qza0YLhBzUKwAFJo6Ab1eEVYfdgEnK4PBMIgBKGJQOIAIQAqgAVbAABwkJAkl1YcFqvHoBHw8BcZHOuCoglYYlE6OksluSlUDyewRer0G-y+4X0v1ZNnsBiBYxBPxkEJWUFW2DwhGIx24YmgrBxdAulMkVnI3HIKP4GFwdFYzC2tXFAEEoNoUrgkbtSkw2BwuLxtdBlVdRNqwLgoBhyqI6HAkX6nLA4ITcJRyGSKa6ykrUulMlSTWaLVb6uJ5cdHSQRC6qSlmM0fVQtTAbvI6WpHs83tzPj0OVYeQ5+VAFv0ujNAsFQoLIgMvepBp0uuyJv3h50myNgQ2h5O-AE5j2OYMuQP-gAWAKFZgMai97w2ACs29IPCCB8bOgAbIDSIoLw359oAOyAuCGYj7jlrzrtAfPluBg7nul5zsep7nt+Y7gTed4PtBXhWLBr7vp+sCIeYq7CpCopxAAkjwnDUHAfo7NAYjmjmCAuGSZRVDUOp6ga0qeoSQSiAoGyGhgHpegcJCnBcuaSOK8IkV6SJpGIxGiBASIIHRSK6ggmZ8HQzSJlCmAAGKCCYTEKhQ9HVGAJDqhgAmiBxZQKVAdHlHQuBgKIrAQHRzo2epmmiTS5YtPShQQK8Pr+n8Ux1kYs4DlOLZtpMMULrM3bxeOXRAfoIGPhMSxgjIIAWCA8kZLcyCgLqdAQAA7gACip8jICAuCGO5LgFSAjnOVoMBYMpYBksgPAmE4hWCE4DA1U5rB4oNw1wIVlTkIo0JOWA3VYLgRAADKSLNYgjSAykaaRjXlLg5TaHKUDtQGggwAA6qQLhBMgAAcAAMhUBhATj3U5SKNQGcDjZ+7V+gAjswIRwJNuDTbge0HU4BCkMgMB0GYhURi8FIAIrME8cBoxj80gHg5SPc91DILYhXo7gpDoqwADCEAEDNKAHPk7UFnA8LneQiNwBEERAA)

**Prettier 1.13.7**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAyjAhvAWwRgwEsBnDcgCwgCcSEIBXAc2oxggwDNSTSsbtDgYANoLgAdKABIAHkgwAKANZwAnkoDkAN1xjmcbQBoM6jQCYk28nEhQAJgDUDR0+c0BmHTGqk6FzdjAEoAbhl0DAB5ACUMAmYxGFJxSSpaJMcMACNRPih8+gz6EggcgCt7GEouCBkFJWUZDE8tDD1g0xa261t7aCDDYxMeix8OvwCh9xlwmQWoKJx8OCJYMkoaUowmNg4uXn4yIQwRNMKGxQxHCCmoVmU1TR19Ye0Qs2erHTsHGdC8yWmAAQgBVAAq2AADhISBJCiUshhWHASDx6AR8PBsoJDrgqIJWGJRAjpHJrrd7o8et9Xl1PrSLNYOn9Bq53iE5hEoJFMCtCMROBBuGJoKxeMVzmSzORuORYcdcHRWMx1jU+RgAIJQDQJXDQkoMXZQFjsYVHMpQM6IslkEhgXDWvIYOhwaFuuywODZXBbIkki6iYqJZKpMmanV6rGGuricUWvhW4O29KJcgkF0KuFXJRU-wPZrWlQWenvMzMmxspwc9wV7y+fyBWuhUbFoutTuljpvOs9LuaPrVgHdYsDjQTbRTZtdHqM9sWAAsOgxzAY1BbHgsAFYV6QeH5N-WNAA2X6keSHrrHgDsvzgumIG4ZbdaHfHy46q-XR-7bV3X77le5Z-hYZ6shewF1hgoGaHerIPk+m5ztyMggCYIAQNCKTQOQyCgMqdAQAA7gACsqCB4SguC6BApCOOhIA5HQuBgOoMBYNCrFEsgPAGHYGGCHYDCkSxrBYrx-FwBhFTkPIIIsWxaJYLgRAADKSJJYgCSAXF0MJyAgOQYDkHhGEeniADq9F+MgAAcAAM5lEXYlksdChkenAwmPoxboAI7MAEcCibg4m4FpOl2AQpDIDAdBGBh5ABnAACKzB3HAkXSSAeA5NZji2UglgYfFuCkAirAAMIQAQEkoKahSMcwdgQrgORUXx2lwAAvj1QA)
```sh
# Options (if any):
--single-quote
--no-bracket-spacing
--parser=scss
```

**Input:**
```scss
// Statement is short enough to fit into one line
$x: (key: 'value', key2:'secondValue', key3: 'thirdValue');
// OR multi line should be fine for short objects too
$x: (
  key: 'value',
  key2:'secondValue',
  key3: 'thirdValue'
);


// Statement is short enough to fit into one line
$x: dothing((key: 'value'), (key2: 'secondValue'));
// BUT Split line should get formatted into a single line
$x: dothing(
  (key: 'value'),
  (key2: 'secondValue')
);

// Statement too long for one line, so split arguments
// Any map short enough to fit on one line it can be represented as single line or multi line
// Any map too long to fit one one line must be split
$x: dothing(
  (key: 'value', key2:'secondValue', key3: 'thirdValue'),
  (
     key: 'value',
     key2:'secondValue',
     key3: 'thirdValue'
  ),
  (key4: 'fourthValue', key5: 'fifthValue', key6: 'sixthValue', key7: 'seventhValue'),
  (
    key4: 'fourthValue',
    key5: 'fifthValue',
    key6: 'sixthValue', 
    key7: 'seventhValue'
  )
);
```

**Output:**
```scss
// Statement is short enough to fit into one line
$x: (
  key: "value",
  key2: "secondValue",
  key3: "thirdValue"
);
// OR multi line should be fine for short objects too
$x: (
  key: "value",
  key2: "secondValue",
  key3: "thirdValue"
);

// Statement is short enough to fit into one line
$x: dothing(
  (
    key: "value"
  ),
  (
    key2: "secondValue"
  )
);
// BUT Split line should get formatted into a single line
$x: dothing(
  (
    key: "value"
  ),
  (
    key2: "secondValue"
  )
);

// Statement too long for one line, so split arguments
// Any map short enough to fit on one line it can be represented as single line or multi line
// Any map too long to fit one one line must be split
$x: dothing(
  (
    key: "value",
    key2: "secondValue",
    key3: "thirdValue"
  ),
  (
    key: "value",
    key2: "secondValue",
    key3: "thirdValue"
  ),
  (
    key4: "fourthValue",
    key5: "fifthValue",
    key6: "sixthValue",
    key7: "seventhValue"
  ),
  (
    key4: "fourthValue",
    key5: "fifthValue",
    key6: "sixthValue",
    key7: "seventhValue"
  )
);
```

**Expected behavior:**

Maps that can fit on one line should be left alone rather than forced into multiple lines

```scss
// Statement is short enough to fit into one line
// Statement is short enough to fit into one line
$x: (key: 'value', key2:'secondValue', key3: 'thirdValue');
// OR multi line should be fine for short objects too
$x: (
  key: 'value',
  key2:'secondValue',
  key3: 'thirdValue'
);


// Statement is short enough to fit into one line
$x: dothing((key: 'value'), (key2: 'secondValue'));
// BUT Split line should get formatted into a single line
$x: dothing((key: 'value'), (key2: 'secondValue'));

// Statement too long for one line, so split arguments
// Any map short enough to fit on one line it can be represented as single line or multi line
// Any map too long to fit one one line must be split
$x: dothing(
  (key: 'value', key2:'secondValue', key3: 'thirdValue'),
  (
     key: 'value',
     key2:'secondValue',
     key3: 'thirdValue'
  ),
  (
    key4: 'fourthValue',
    key5: 'fifthValue',
    key6: 'sixthValue',
    key7: 'seventhValue'
  ),
  (
    key4: 'fourthValue',
    key5: 'fifthValue',
    key6: 'sixthValue', 
    key7: 'seventhValue'
  )
);
```

**Comment by evilebottnawi** [comment]
@BPScott @j-f1 We print in one line before, but then we decided to always insert newline for better reading, output expected, it is not the bug. SCSS map nearer JSON objects than JS objects.

**Comment by BPScott** [comment]
Fair enough :)

Personally I think using having logic similar to JS objects is preferable to JSON-style objects :) I use short maps that would fit onto a single line if not for the current logic often enough to notice the splitting behaviour.

**Comment by BPScott** [comment]
Fair enough :)

Personally I think using having logic similar to JS objects is preferable to JSON-style objects :) I use short maps that would fit onto a single line if not for the current logic often enough to notice the splitting behaviour.

**Comment by evilebottnawi** [comment]
@BPScott So we can close issue? Also it is better to add new value to:
```scss
$point: (
  x: 123
);
```

You don't need add newlines before `(` and after `)`.

**Comment by rowild** [comment]
This cannot be closed yet as things like this should be supported:

$ratio: (
  "4\\:3": (4: 3),
);

Ratio-dependent image formatting is pretty common nowadays and should receive better styling-care.

**Comment by evilebottnawi** [comment]
@rowild Looks very unreasonable:
```
$ratio: (
"4\:3": (4: 3),
);
```

This is ugly, when you need add new key/value you need add newlines and it is very unconventional.

Better:
```
$ratio: (
  "4\:3": (
    4: 3
  )
);

```

Just add new value and all, also you can enable trailing commas and you diff will be very small an d readable.

**Comment by rowild** [comment]
Totally unacceptable, when you have to manage of 20+ ratios. Just not good reading nor understanding possible.

**Comment by rowild** [comment]
Prettier should at least offer a flag to handle such a situation. But I guess that's against the project's intention...

**Comment by evilebottnawi** [comment]
> Totally unacceptable, when you have to manage of 20+ ratios. Just not good reading nor understanding possible.

Please provide example

> Prettier should at least offer a flag to handle such a situation. But I guess that's against the project's intention...

Yep

**Comment by rowild** [comment]
Example? It's easy to reproduce by copying and pasting... C'mon, man!

**Comment by evilebottnawi** [comment]
@BPScott Please provide example where it look bad and unreadable?

**Comment by j-f1** [comment]
@evilebottnawi I agree that the breaking here is unnecessary. Is it possible to format the same way as JS objects?

**Comment by BPScott** [comment]
@rowild, "just copy and paste" when asked for examples is almost certainly not going to make the team feel engaged and helpful; it's probably going to do the exact opposite. If we want to take up @evilebottnawi's valuable time with this issue I'd suggest giving him as much context as possible so that he can empathize with your concerns.


Heya @evilebottnawi, sorry for missing your original comment. I think this is worth keeping open for further discussion (hey look we're having it now). I totally agree that multi-line maps (with trailing-commas) are preferable with longer lists for the sake of easier diffs and that is how I prefer to write code even without prettier's assistance, however for shorter lists - that I am confident won't change - there is potential for creating lots of whitespace.  My main concern stems from two cases that result in what I feel is excessive whitespace that could be wrote in a more compact style - and the babylon parser uses that more compact style if you to create similar constructs in JS.

My original issue contained abstract code for the sake of completeness. Here's some examples taken from my work code base.

### Calling a function with a map

I think the babylon parser's approach of opening the object on the same line as the function declaration to reduce nesting and whitespace is more readable:

The following is considered formatted by prettier:

```scss
$map: map-merge(
  $map,
  (
    $key: $value,
  )
);

$color-palette-data: map-extend(
  $color-palette-data,
  (
    'state': (
      'hover': rgba(223, 227, 232, 0.3),
      'focused': rgba(223, 227, 232, 0.3),
      'active': rgba(179, 188, 245, 0.1),
      'selected': rgba(179, 188, 245, 0.15),
      'subdued': rgba(249, 250, 251, 1),
      'disabled': rgba(249, 250, 251, 1),
      'hover-destructive': rgba(251, 234, 229, 0.4),
      'focused-destructive': rgba(251, 234, 229, 0.4),
      'active-destructive': rgba(220, 56, 37, 0.03),
    ),
  )
);
```

I think this would be cleaner to write as:

```scss
$map: map-merge($map, ($key: $value));

$color-palette-data: map-extend($color-palette-data, (
  'state': (
    'hover': rgba(223, 227, 232, 0.3),
    'focused': rgba(223, 227, 232, 0.3),
    'active': rgba(179, 188, 245, 0.1),
    'selected': rgba(179, 188, 245, 0.15),
    'subdued': rgba(249, 250, 251, 1),
    'disabled': rgba(249, 250, 251, 1),
    'hover-destructive': rgba(251, 234, 229, 0.4),
    'focused-destructive': rgba(251, 234, 229, 0.4),
    'active-destructive': rgba(220, 56, 37, 0.03),
  ),
));
```

Admittedly the map-extend example is a minor annoyance but the map-merge example shows a large amount if whitespace that IMO hinders readability (1 line becomes 6).


### Nested maps

In cases where a large map whose values are small maps it can be hard to follow the top-level map's keys as they are so spread out.

Prettier considers this as formatted:

```scss
$font-size-data: (
  caption: (
    base: rem(13px),
    large-screen: rem(12px),
  ),
  heading: (
    base: rem(17px),
    large-screen: rem(16px),
  ),
  subheading: (
    base: rem(13px),
    large-screen: rem(12px),
  ),
  input: (
    base: rem(16px),
    large-screen: rem(14px),
  ),
  body: (
    base: rem(15px),
    large-screen: rem(14px),
  ),
  button: (
    base: rem(15px),
    large-screen: rem(14px),
  ),
  button-large: (
    base: rem(17px),
    large-screen: rem(16px),
  ),
  display-x-large: (
    base: rem(27px),
    large-screen: rem(42px),
  ),
  display-large: (
    base: rem(24px),
    large-screen: rem(28px),
  ),
  display-medium: (
    base: rem(21px),
    large-screen: rem(26px),
  ),
  display-small: (
    base: rem(16px),
    large-screen: rem(20px),
  ),
);
```

I think this would be cleaner to write as:

```scss
$font-size-data: (
  caption: (base: rem(13px), large-screen: rem(12px)),
  heading: (base: rem(17px), large-screen: rem(16px)),
  subheading: (base: rem(13px), large-screen: rem(12px)),
  input: (base: rem(16px), large-screen: rem(14px)),
  body: (base: rem(15px), large-screen: rem(14px)),
  button: (base: rem(15px), large-screen: rem(14px)),
  button-large: (base: rem(17px), large-screen: rem(16px)),
  display-x-large: (base: rem(27px), large-screen: rem(42px)),
  display-large: (base: rem(24px), large-screen: rem(28px)),
  display-medium: (base: rem(21px), large-screen: rem(26px)),
  display-small: (base: rem(16px), large-screen: rem(20px)),
);
```

**Comment by BPScott** [comment]
I imagine @rowild's ratio concerns are similar to my nested maps concern. I reckon he's got one big map of multiple ratios that is formatted like so:

```scss
$ratio: (
  "1\:1": (
    1: 1
  ),
  "2\:1": (
    2: 1
  ),
  "3\:2": (
    3: 2
  ),
  "4\:3": (
    4: 3
  ),
  "16\:9": (
    16: 9
  )
);
```

Which would look cleaner as:

```scss
$ratio: (
  "1\:1": (1:1),
  "2\:1": (2: 1),
  "3\:2": (3: 2),
  "4\:3": (4: 3),
  "16\:9": (16:9),
);
```

(Part of me wonders why that needs a map and thinks having single number - `4/3` instead of `(4:3)` would be clearer but that's a different problem)

**Comment by rowild** [comment]
@BPScott I so do not agree with you! As you very well know: all open source projects have agreed on pretty much these 2 things:
- Do the research.
- Post the shortest possible code to reproduce the scenario.
And in this case: use the playground

As you also very well know, this can take hours. So when I eventually follow those rules and then the project owner comes up with something that contradicts them - and only, because he himself does not want to test it -, I really feel pranked and not taken seriously, since I just did my job. (And it is not the only project, where research has to be done... so time sums up). Now I expect the other party to do his/her job. You also have to stick to those rules.

Furthermore: you get one result using this:
```scss
$ratio: (
  "1\:1": "1:1",
  "2\:1": "2: 1",
  "3\:2": 3: 2,
  "4\:3": 4: 3,
  "16\:9": 16:9,
);
```
(result on one line), and another one as soon as parenthesis are used. Probably something what you'd call an edge case - but edge cases need to be handled and respected as well, only then the quality of a(gain another) project becomes clear.

There must be place for critique. If not, then don't make it open source.

As a side note: I **must** use prettier (in office, not private), but I do not like it. Nevertheless I contribute by reporting and giving feedback. And @evilebottnawi: answering with "Feel free to contribute code" - as you did today in another post - is just not an answer, not in generell and not to my question specifically (see the comment in another post)! This is just wasting time, which is unfair, because we all have our time issues, not only you. We are on github here - it should be clear without saying that contributing code is pretty much one of the main reasons why there is github... right? So please provide substantial answers - or none at all. Everything else and a waste of time, yours and mine.
Thanks.
